utils/form_validation.php

```diff
@@ -62,13 +70,35 @@ function get_field(string $key): Field
 
     function render_errors()
     {
-        $errors = "";
+        $result = "";
+
+        // Add form action name
+        if ($this->action)
+            $result .= "<input type=\"hidden\" name=\"action\" value=\"$this->action\">";
+
+        // Add csrf
+        $result .= set_csrf();
+
         foreach ($this->fields as $field) {
-            if ($field->error)
-                $errors .= "<label for=\"$field->key\" class=\"error\">{$field->get_label()} : $field->error</label>";
+            if ($field->error) {
+                $result .= "<label for=\"$field->key\" class=\"error\">{$field->get_label()} : $field->error</label>";
+            }
+        }
+        if (!$this->empty and !$this->valid()) {
+            return $result . "<br/><br/>";
         }
-        if ($errors != "") {
-            return $errors . "<br/><br/>";
+        return $result;
+    }
+
+    function set_success($success)
+    {
+        $this->success = $success;
+    }
+
+    function render_success()
```

Comment:
Maybe consider merging this with `render_errors()` into a function like `render_meta()` or something